### ISSUE TITLE
s3_client.copy not working

### ISSUE THREAD

**Issue by leezu** [terrycain/aioboto3#68]
* Async AWS SDK for Python version:
```
Requirement already satisfied: aioboto3 in ./anaconda3/envs/py36/lib/python3.6/site-packages (4.0.1)
Requirement already satisfied: botocore<=1.10.12 in ./anaconda3/envs/py36/lib/python3.6/site-packages (from aioboto3) (1.10.12)
Requirement already satisfied: boto3<=1.7.12,>=1.4.7 in ./anaconda3/envs/py36/lib/python3.6/site-packages (from aioboto3) (1.7.12)
Requirement already satisfied: aiobotocore==0.8.0 in ./anaconda3/envs/py36/lib/python3.6/site-packages (from aioboto3) (0.8.0)
Requirement already satisfied: jmespath<1.0.0,>=0.7.1 in ./anaconda3/envs/py36/lib/python3.6/site-packages (from botocore<=1.10.12->aioboto3) (0.9.3)
Requirement already satisfied: docutils>=0.10 in ./anaconda3/envs/py36/lib/python3.6/site-packages (from botocore<=1.10.12->aioboto3) (0.14)
Requirement already satisfied: python-dateutil<3.0.0,>=2.1; python_version >= "2.7" in ./anaconda3/envs/py36/lib/python3.6/site-packages (from botocore<=1.10.12->aioboto3) (2.7.3)
Requirement already satisfied: s3transfer<0.2.0,>=0.1.10 in ./anaconda3/envs/py36/lib/python3.6/site-packages (from boto3<=1.7.12,>=1.4.7->aioboto3) (0.1.13)
Requirement already satisfied: wrapt>=1.10.10 in ./anaconda3/envs/py36/lib/python3.6/site-packages (from aiobotocore==0.8.0->aioboto3) (1.10.11)
Requirement already satisfied: packaging>=16.8 in ./anaconda3/envs/py36/lib/python3.6/site-packages (from aiobotocore==0.8.0->aioboto3) (17.1)
Requirement already satisfied: aiohttp<3.2.0,>=3.1.0 in ./anaconda3/envs/py36/lib/python3.6/site-packages (from aiobotocore==0.8.0->aioboto3) (3.1.3)
Requirement already satisfied: multidict>=2.1.4 in ./anaconda3/envs/py36/lib/python3.6/site-packages (from aiobotocore==0.8.0->aioboto3) (4.3.1)
Requirement already satisfied: six>=1.5 in ./anaconda3/envs/py36/lib/python3.6/site-packages (from python-dateutil<3.0.0,>=2.1; python_version >= "2.7"->botocore<=1.10.12->aioboto3) (1.11.0)
Requirement already satisfied: pyparsing>=2.0.2 in ./anaconda3/envs/py36/lib/python3.6/site-packages (from packaging>=16.8->aiobotocore==0.8.0->aioboto3) (2.2.0)
Requirement already satisfied: yarl<2.0,>=1.0 in ./anaconda3/envs/py36/lib/python3.6/site-packages (from aiohttp<3.2.0,>=3.1.0->aiobotocore==0.8.0->aioboto3) (1.2.6)
Requirement already satisfied: idna-ssl>=1.0 in ./anaconda3/envs/py36/lib/python3.6/site-packages (from aiohttp<3.2.0,>=3.1.0->aiobotocore==0.8.0->aioboto3) (1.0.1)
Requirement already satisfied: async-timeout<3.0,>=1.2 in ./anaconda3/envs/py36/lib/python3.6/site-packages (from aiohttp<3.2.0,>=3.1.0->aiobotocore==0.8.0->aioboto3) (2.0.1)
Requirement already satisfied: attrs>=17.3.0 in ./anaconda3/envs/py36/lib/python3.6/site-packages (from aiohttp<3.2.0,>=3.1.0->aiobotocore==0.8.0->aioboto3) (18.1.0)
Requirement already satisfied: chardet<4.0,>=2.0 in ./.local/lib/python3.6/site-packages (from aiohttp<3.2.0,>=3.1.0->aiobotocore==0.8.0->aioboto3) (3.0.4)
Requirement already satisfied: idna>=2.0 in ./.local/lib/python3.6/site-packages (from yarl<2.0,>=1.0->aiohttp<3.2.0,>=3.1.0->aiobotocore==0.8.0->aioboto3) (2.6)

```
* Python version: Py3.6.6
* Operating System: OS X

### Description

I try to use s3_client.copy for copying large files (> 5GB), requiring the multi-part support of the copy method. Unfortunately that fails.

### What I Did

```
async def copy_rename(s3, name):
    bucket = A
    sha1sum_key = B

    # get object from s3
    response = await s3.get_object(Bucket=bucket, Key=sha1sum_key)
    async with response['Body'] as stream:
        sha1sum = (await stream.read()).decode('ascii').strip()

    response = await s3.copy(
        Bucket=C,
        Key=f'{name}-{sha1sum[:8]}.bin',
        CopySource=dict(Bucket=D, Key=f'{name}.bin'))


async def go():
    async with aioboto3.client('s3', region_name='us-east-1') as s3:
        tasks = []
        for name in NAMES:
            tasks.append(copy_rename(s3, name))
        await asyncio.gather(*tasks)

```

Fails with

```
Traceback (most recent call last):
  File "s3_links.py", line 33, in <module>
    loop.run_until_complete(go())
  File "/Users/lllausen/anaconda3/envs/py36/lib/python3.6/asyncio/base_events.py", line 468, in run_until_complete
    return future.result()
  File "s3_links.py", line 29, in go
    await asyncio.gather(*tasks)
  File "s3_links.py", line 21, in copy_rename
    Bucket=..., Key=...
  File "/Users/lllausen/anaconda3/envs/py36/lib/python3.6/site-packages/boto3/s3/inject.py", line 271, in copy
    return future.result()
  File "/Users/lllausen/anaconda3/envs/py36/lib/python3.6/site-packages/s3transfer/futures.py", line 73, in result
    return self._coordinator.result()
  File "/Users/lllausen/anaconda3/envs/py36/lib/python3.6/site-packages/s3transfer/futures.py", line 233, in result
    raise self._exception
  File "/Users/lllausen/anaconda3/envs/py36/lib/python3.6/site-packages/s3transfer/tasks.py", line 255, in _main
    self._submit(transfer_future=transfer_future, **kwargs)
  File "/Users/lllausen/anaconda3/envs/py36/lib/python3.6/site-packages/s3transfer/copies.py", line 112, in _submit
    response['ContentLength'])
TypeError: 'coroutine' object is not subscriptable

```

**Comment by terrycain** [comment]
Ahh of course, s3transfer appears again.

So s3.download_file/upload_file and their derivatives make use of the `s3transfer` module which in turn does multipart uploads. I've patched upload/download file with hacky non-multipart workarounds thus far. 

I have a branch `multipart_upload_60` which has mulitpart uploads, so I reckon i can repurpose most of that and stream the download along with uploading many chunks.

Give me a couple of days and we'll see what I can come up with. It might take a few iterations as I don't use copy on a day-to-day basis so will only throw it through a few basic tests.

**Comment by terrycain** [comment]
@leezu can you try out the code on branch `multipart_upload_60` I added a nice (mostly copy and paste) `copy` function in 63abdd5e21d897b54bbcae4b35c5e0abae27c73a which uses multipart.

It works for my super simple test case of copying `b'Hello World\n'` but would be nice for that code to have a decent real-life test. Also I have no clue what errors `s3transfer` would return so currently it returns some roughly sane errors but i'll look through `s3transfer`'s code and see if there's anything better to error with.

**Comment by terrycain** [comment]
Is most likely fixed in v5.0.0
